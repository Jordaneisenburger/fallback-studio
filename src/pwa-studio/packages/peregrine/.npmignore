.eslintrc.js
/lib/**/__{docs,helpers,mocks,tests}__/**
/scripts
