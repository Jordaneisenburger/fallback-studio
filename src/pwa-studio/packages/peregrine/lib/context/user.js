import React, { createContext, useContext, useMemo } from 'react';
import { connect } from 'react-redux';

import actions from '../store/actions/user/actions';
import * as asyncActions from '../store/actions/user/asyncActions';
import bindActionCreators from '../util/bindActionCreators';

const UserContext = createContext();

const UserContextProvider = props => {
    const { actions, asyncActions, children, userState } = props;

    const userApi = useMemo(
        () => ({
            actions,
            ...asyncActions
        }),
        [actions, asyncActions]
    );

    const contextValue = useMemo(() => [userState, userApi], [
        userApi,
        userState
    ]);

    return (
        <UserContext.Provider value={contextValue}>
            {children}
        </UserContext.Provider>
    );
};

const mapStateToProps = ({ user }) => ({ userState: user });

const mapDispatchToProps = dispatch => ({
    actions: bindActionCreators(actions, dispatch),
    asyncActions: bindActionCreators(asyncActions, dispatch)
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(UserContextProvider);

export const useUserContext = () => useContext(UserContext);
