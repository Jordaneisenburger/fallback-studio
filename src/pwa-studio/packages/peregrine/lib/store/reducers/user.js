import { handleActions } from 'redux-actions';

import { Util } from '../../index';
const { BrowserPersistence } = Util;

const storage = new BrowserPersistence();

import actions from '../actions/user';

export const name = 'user';

const isSignedIn = () => !!storage.getItem('signin_token');

const initialState = {
    currentUser: {
        email: '',
        firstname: '',
        lastname: ''
    },
    getDetailsError: null,
    isGettingDetails: false,
    isResettingPassword: false,
    isSignedIn: isSignedIn(),
    isSigningIn: false,
    resetPasswordError: null,
    signInError: null
};

const reducerMap = {
    [actions.signIn.request]: state => {
        return {
            ...state,
            isSigningIn: true,
            signInError: null
        };
    },
    [actions.signIn.receive]: (state, { payload, error }) => {
        if (error) {
            return {
                ...initialState,
                signInError: payload
            };
        }

        return {
            ...state,
            isSignedIn: true,
            isSigningIn: false,
            signInError: null
        };
    },
    [actions.getDetails.request]: state => {
        return {
            ...state,
            getDetailsError: null,
            isGettingDetails: true
        };
    },
    [actions.getDetails.receive]: (state, { payload, error }) => {
        if (error) {
            return {
                ...state,
                getDetailsError: payload,
                isGettingDetails: false
            };
        }

        return {
            ...state,
            currentUser: payload,
            getDetailsError: null,
            isGettingDetails: false
        };
    },
    [actions.resetPassword.request]: state => ({
        ...state,
        isResettingPassword: true
    }),
    // TODO: handle the reset password response from the API.
    [actions.resetPassword.receive]: (state, { payload, error }) => {
        if (error) {
            return {
                ...state,
                isResettingPassword: false,
                resetPasswordError: payload
            };
        }

        return {
            ...state,
            isResettingPassword: false,
            resetPasswordError: null
        };
    },
    [actions.reset]: () => {
        return {
            ...initialState,
            isSignedIn: isSignedIn()
        };
    }
};

export default handleActions(reducerMap, initialState);
