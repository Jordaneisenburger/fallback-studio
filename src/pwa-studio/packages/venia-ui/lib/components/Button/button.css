.root {
    composes: root from '../clickable.css';
    background: none;
    border: 1px solid rgb(var(--color));
    border-radius: 1.5rem;
    color: rgb(var(--color));
    font-size: 0.75rem;
    font-weight: 600;
    height: 2rem;
    min-width: 7.5rem;
    padding: 0 1rem;
    transition-duration: 384ms;
    transition-property: background-color, color;
    transition-timing-function: var(--venia-anim-standard);
    --color: var(--venia-text);
}

.filled {
    background-color: rgb(var(--color));
    color: white;
}

.root:hover {
    --color: var(--venia-teal);
}

.root:focus {
    box-shadow: 0 0 0 2px rgb(var(--venia-teal-light)),
        0 0 0.5rem 2px rgba(var(--venia-teal), 0.2);
    outline: none;
    transition-duration: 128ms;
    --color: var(--venia-teal);
}

.root:active {
    transition-duration: 128ms;
    --color: var(--venia-teal-dark);
}

.root:disabled {
    pointer-events: none;
    --color: var(--venia-grey-dark);
}

.content {
    align-items: center;
    display: inline-grid;
    gap: 0.5rem;
    grid-auto-flow: column;
    justify-content: center;
    justify-items: center;
}

/* derived classes */

.root_lowPriority {
    composes: root;
    border-width: 0;
    text-decoration: underline;
}

.root_lowPriority:focus {
    box-shadow: none;
}

.root_normalPriority {
    composes: root;
}
.root_highPriority {
    composes: root filled;
}
