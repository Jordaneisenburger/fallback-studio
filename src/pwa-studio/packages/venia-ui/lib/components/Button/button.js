import React from 'react';
import { oneOf, shape, string } from 'prop-types';

import { mergeClasses } from '../../classify';
import defaultClasses from './button.css';

const getRootClassName = priority => `root_${priority}Priority`;

/**
 * A component for buttons.
 *
 * @typedef Button
 * @kind functional component
 *
 * @param {props} props React component props
 *
 * @returns {React.Element} A React component that displays a single button.
 */
const Button = props => {
    const {
        children,
        classes: propClasses,
        priority,
        type,
        ...restProps
    } = props;
    const classes = mergeClasses(defaultClasses, propClasses);
    const rootClassName = classes[getRootClassName(priority)];

    return (
        <button className={rootClassName} type={type} {...restProps}>
            <span className={classes.content}>{children}</span>
        </button>
    );
};

/**
 * Props for {@link Button}
 *
 * @typedef props
 *
 * @property {Object} classes An object containing the class names for the
 * Button component.
 * @property {string} classes.content classes for the button content
 * @property {string} classes.root classes for root container
 * @property {string} classes.root_highPriority classes for Button if
 * high priority.
 * @property {string} classes.root_lowPriority classes for Button if
 * low priority.
 * @property {string} classes.root_normalPriority classes for Button if
 * normal priority.
 * @property {string} priority the priority/importance of the Button
 * @property {string} type the type of the Button
 */
Button.propTypes = {
    classes: shape({
        content: string,
        root: string,
        root_highPriority: string,
        root_lowPriority: string,
        root_normalPriority: string
    }),
    priority: oneOf(['high', 'low', 'normal']).isRequired,
    type: oneOf(['button', 'reset', 'submit']).isRequired
};

Button.defaultProps = {
    priority: 'normal',
    type: 'button'
};

export default Button;
