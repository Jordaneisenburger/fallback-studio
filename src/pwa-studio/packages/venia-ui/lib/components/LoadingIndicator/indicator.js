import React from 'react';

import defaultClasses from './indicator.css';
import { mergeClasses } from '../../classify';
import Image from '../Image';

import logo from '../Logo/logo.svg';

const LoadingIndicator = props => {
    const classes = mergeClasses(defaultClasses, props.classes);
    const className = props.global ? classes.global : classes.root;

    return (
        <div className={className}>
            <Image
                alt="Loading indicator"
                classes={{ image: classes.indicator }}
                displayPlaceholder={false}
                height="64"
                src={logo}
                width="64"
            />
            <span className={classes.message}>{props.children}</span>
        </div>
    );
};

export default LoadingIndicator;
