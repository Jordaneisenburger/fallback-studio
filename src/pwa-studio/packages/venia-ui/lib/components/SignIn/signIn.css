.root {
    display: grid;
    gap: 1.5rem;
    justify-items: stretch;
    padding: 1rem 1.5rem;
}

.forgotPassword {
    display: block;
    text-decoration: underline;
    color: rgb(var(--venia-error));
    font-size: 0.8rem;
}

.form {
    display: grid;
    row-gap: 1rem;
}

.modal {
    visibility: hidden;
    height: 100%;
    width: 100%;
    background-color: rgb(var(--venia-grey));
    text-align: center;
    position: absolute;
    bottom: 0;
}

.modal_active {
    visibility: visible;
    composes: modal;
    opacity: 0.9;
}

.createAccountButton,
.forgotPasswordButton,
.signInButton {
    align-items: center;
    display: flex;
    justify-content: center;
    text-align: center;
}

.forgotPasswordButtonRoot {
    composes: root_lowPriority from '../Button/button.css';
    color: rgb(var(--venia-error));
}

.signInButton {
    margin-top: 1rem;
}

.signInDivider {
    border-color: rgb(var(--venia-border));
    border-style: solid;
    border-width: 0 0 1px;
    margin: 1.5rem 1rem;
}

.signInError {
    background-color: rgb(var(--venia-grey));
    border-radius: 4px;
    color: rgb(var(--venia-error));
    font-size: 0.875rem;
    line-height: 1.25rem;
    padding: 1rem;
}

.signInError:empty {
    display: none;
}
