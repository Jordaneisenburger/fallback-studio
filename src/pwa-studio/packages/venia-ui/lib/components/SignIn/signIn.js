import React from 'react';
import { func, shape, string } from 'prop-types';
import { Form } from 'informed';

import { mergeClasses } from '../../classify';
import Button from '../Button';
import Field from '../Field';
import LoadingIndicator from '../LoadingIndicator';
import TextInput from '../TextInput';
import { isRequired, validateEmail } from '../../util/formValidators';
import combine from '../../util/combineValidators';

import defaultClasses from './signIn.css';
import { useSignIn } from '@magento/peregrine/lib/talons/SignIn/useSignIn';
import CREATE_CART_MUTATION from '../../queries/createCart.graphql';
import GET_CUSTOMER_QUERY from '../../queries/getCustomer.graphql';
import SIGN_IN_MUTATION from '../../queries/signIn.graphql';
import GET_CART_DETAILS_QUERY from '../../queries/getCartDetails.graphql';

const SignIn = props => {
    const classes = mergeClasses(defaultClasses, props.classes);
    const { setDefaultUsername, showCreateAccount, showForgotPassword } = props;

    const talonProps = useSignIn({
        createCartMutation: CREATE_CART_MUTATION,
        customerQuery: GET_CUSTOMER_QUERY,
        getCartDetailsQuery: GET_CART_DETAILS_QUERY,
        signInMutation: SIGN_IN_MUTATION,
        setDefaultUsername,
        showCreateAccount,
        showForgotPassword
    });

    const {
        errors,
        formRef,
        handleCreateAccount,
        handleForgotPassword,
        handleSubmit,
        isBusy
    } = talonProps;

    // Map over any errors we get and display an appropriate error.
    const errorMessage = errors.length
        ? errors
              .map(({ message }) => message)
              .reduce((acc, msg) => msg + '\n' + acc, '')
        : null;

    if (isBusy) {
        return (
            <div className={classes.modal_active}>
                <LoadingIndicator>{'Signing In'}</LoadingIndicator>
            </div>
        );
    }

    return (
        <div className={classes.root}>
            <Form
                ref={formRef}
                className={classes.form}
                onSubmit={handleSubmit}
            >
                <Field label="Email" required={true}>
                    <TextInput
                        autoComplete="email"
                        field="email"
                        validate={combine([isRequired, validateEmail])}
                    />
                </Field>
                <Field label="Password" required={true}>
                    <TextInput
                        autoComplete="current-password"
                        field="password"
                        type="password"
                        validate={isRequired}
                    />
                </Field>
                <div className={classes.signInError}>{errorMessage}</div>
                <div className={classes.signInButton}>
                    <Button priority="high" type="submit">
                        {'Sign In'}
                    </Button>
                </div>
            </Form>
            <div className={classes.forgotPasswordButton}>
                <Button
                    priority="low"
                    type="button"
                    onClick={handleForgotPassword}
                    classes={{
                        root_lowPriority: classes.forgotPasswordButtonRoot
                    }}
                >
                    {'Forgot Password?'}
                </Button>
            </div>
            <div className={classes.signInDivider} />
            <div className={classes.createAccountButton}>
                <Button
                    priority="normal"
                    type="button"
                    onClick={handleCreateAccount}
                >
                    {'Create an Account'}
                </Button>
            </div>
        </div>
    );
};

export default SignIn;

SignIn.propTypes = {
    classes: shape({
        createAccountButton: string,
        form: string,
        forgotPasswordButton: string,
        forgotPasswordButtonRoot: string,
        root: string,
        signInButton: string,
        signInDivider: string,
        signInError: string
    }),
    setDefaultUsername: func.isRequired,
    showCreateAccount: func.isRequired,
    showForgotPassword: func.isRequired
};
