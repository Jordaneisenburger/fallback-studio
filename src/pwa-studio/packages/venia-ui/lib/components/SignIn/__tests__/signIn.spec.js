import React from 'react';
import { act } from 'react-test-renderer';
import { Form } from 'informed';
import { createTestInstance } from '@magento/peregrine';

import Button from '../../Button';
import LoadingIndicator from '../../LoadingIndicator';
import SignIn from '../signIn';
import { useUserContext } from '@magento/peregrine/lib/context/user';

jest.mock('../../../classify');
jest.mock('../../Button', () => () => <i />);
jest.mock('../../LoadingIndicator', () => () => <i />);

jest.mock('@magento/peregrine/lib/context/user', () => {
    const userState = {
        isGettingDetails: false,
        isSigningIn: false,
        signInError: null,
        getDetailsError: null
    };
    const userApi = { signIn: jest.fn() };
    const useUserContext = jest.fn(() => [userState, userApi]);

    return { useUserContext };
});

const props = {
    setDefaultUsername: jest.fn(),
    showCreateAccount: jest.fn(),
    showForgotPassword: jest.fn(),
    signIn: jest.fn(),
    hasError: false,
    isSigningIn: false
};

test('renders correctly', () => {
    const component = createTestInstance(<SignIn {...props} />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('renders the loading indicator when form is submitting', () => {
    const [userState, userApi] = useUserContext();
    useUserContext.mockReturnValueOnce([
        { ...userState, isSigningIn: true },
        userApi
    ]);

    const testProps = {
        ...props
    };
    const { root } = createTestInstance(<SignIn {...testProps} />);

    act(() => {
        expect(root.findByType(LoadingIndicator)).toBeTruthy();
    });
});

test('displays an error message if there is a sign in error', () => {
    const [userState, userApi] = useUserContext();
    useUserContext.mockReturnValueOnce([
        { ...userState, signInError: new Error() },
        userApi
    ]);
    const testProps = {
        ...props
    };

    const component = createTestInstance(<SignIn {...testProps} />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('calls `signIn` on submit', () => {
    const [, { signIn }] = useUserContext();

    const values = { email: 'a', password: 'b' };

    const { root } = createTestInstance(<SignIn {...props} />);

    act(() => {
        root.findByType(Form).props.onSubmit(values);
    });

    expect(signIn).toHaveBeenCalledTimes(1);
    expect(signIn).toHaveBeenNthCalledWith(1, {
        username: values.email,
        password: values.password
    });
});

test('changes view to CreateAccount', () => {
    const { setDefaultUsername, showCreateAccount } = props;
    const { root } = createTestInstance(<SignIn {...props} />);

    const { onClick } = root
        .findByProps({ className: 'createAccountButton' })
        .findByType(Button).props;

    act(() => {
        onClick();
    });

    expect(setDefaultUsername).toHaveBeenCalledTimes(1);
    expect(showCreateAccount).toHaveBeenCalledTimes(1);
});

test('changes view to ForgotPassword', () => {
    const { setDefaultUsername, showForgotPassword } = props;
    const { root } = createTestInstance(<SignIn {...props} />);

    const { onClick } = root
        .findByProps({ className: 'forgotPasswordButton' })
        .findByType(Button).props;

    act(() => {
        onClick();
    });

    expect(setDefaultUsername).toHaveBeenCalledTimes(1);
    expect(showForgotPassword).toHaveBeenCalledTimes(1);
});

test("avoids reading the form if it doesn't exist", () => {
    const { setDefaultUsername, showCreateAccount, showForgotPassword } = props;
    const instance = createTestInstance(<SignIn {...props} />);
    const { root } = instance;

    const { onClick: createAccount } = root
        .findByProps({ className: 'createAccountButton' })
        .findByType(Button).props;
    const { onClick: forgotPassword } = root
        .findByProps({ className: 'forgotPasswordButton' })
        .findByType(Button).props;

    instance.unmount();

    act(() => {
        createAccount();
    });

    act(() => {
        forgotPassword();
    });

    expect(setDefaultUsername).not.toHaveBeenCalled();
    expect(showCreateAccount).toHaveBeenCalledTimes(1);
    expect(showForgotPassword).toHaveBeenCalledTimes(1);
});
