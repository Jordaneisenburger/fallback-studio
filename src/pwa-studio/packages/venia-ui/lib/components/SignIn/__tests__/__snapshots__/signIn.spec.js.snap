// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays an error message if there is a sign in error 1`] = `
<div
  className="root"
>
  <form
    className="form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="root"
    >
      <label
        className="label"
      >
        <span
          className="requiredSymbol"
        />
        Email
      </label>
      <span
        className="root"
        style={
          Object {
            "--iconsAfter": 0,
            "--iconsBefore": 0,
          }
        }
      >
        <span
          className="input"
        >
          <input
            autoComplete="email"
            className="input"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            value=""
          />
        </span>
        <span
          className="before"
        />
        <span
          className="after"
        />
      </span>
      <p
        className="root"
      />
    </div>
    <div
      className="root"
    >
      <label
        className="label"
      >
        <span
          className="requiredSymbol"
        />
        Password
      </label>
      <span
        className="root"
        style={
          Object {
            "--iconsAfter": 0,
            "--iconsBefore": 0,
          }
        }
      >
        <span
          className="input"
        >
          <input
            autoComplete="current-password"
            className="input"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
          />
        </span>
        <span
          className="before"
        />
        <span
          className="after"
        />
      </span>
      <p
        className="root"
      />
    </div>
    <div
      className="signInError"
    >
      The account sign-in was incorrect or your account is disabled temporarily. Please wait and try again later.

    </div>
    <div
      className="signInButton"
    >
      <i />
    </div>
  </form>
  <div
    className="forgotPasswordButton"
  >
    <i />
  </div>
  <div
    className="signInDivider"
  />
  <div
    className="createAccountButton"
  >
    <i />
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="root"
>
  <form
    className="form"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="root"
    >
      <label
        className="label"
      >
        <span
          className="requiredSymbol"
        />
        Email
      </label>
      <span
        className="root"
        style={
          Object {
            "--iconsAfter": 0,
            "--iconsBefore": 0,
          }
        }
      >
        <span
          className="input"
        >
          <input
            autoComplete="email"
            className="input"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            value=""
          />
        </span>
        <span
          className="before"
        />
        <span
          className="after"
        />
      </span>
      <p
        className="root"
      />
    </div>
    <div
      className="root"
    >
      <label
        className="label"
      >
        <span
          className="requiredSymbol"
        />
        Password
      </label>
      <span
        className="root"
        style={
          Object {
            "--iconsAfter": 0,
            "--iconsBefore": 0,
          }
        }
      >
        <span
          className="input"
        >
          <input
            autoComplete="current-password"
            className="input"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value=""
          />
        </span>
        <span
          className="before"
        />
        <span
          className="after"
        />
      </span>
      <p
        className="root"
      />
    </div>
    <div
      className="signInError"
    />
    <div
      className="signInButton"
    >
      <i />
    </div>
  </form>
  <div
    className="forgotPasswordButton"
  >
    <i />
  </div>
  <div
    className="signInDivider"
  />
  <div
    className="createAccountButton"
  >
    <i />
  </div>
</div>
`;
